test/e2e/network/loadbalancer.go

```diff
@@ -299,21 +299,21 @@ var _ = common.SIGDescribe("LoadBalancers", func() {
 		framework.Logf("service port TCP: %d", svcPort)
 
 		ginkgo.By("hitting the TCP service's LoadBalancer")
-		e2eservice.TestReachableHTTP(ctx, tcpIngressIP, svcPort, loadBalancerCreateTimeout)
+		e2eservice.TestReachableHTTP(ctx, tcpIngressIP, svcPort, loadBalancerLagTimeout)
 
 		ginkgo.By("Scaling the pods to 0")
 		err = tcpJig.Scale(ctx, 0)
 		framework.ExpectNoError(err)
 
-		ginkgo.By("looking for ICMP REJECT on the TCP service's LoadBalancer")
-		testRejectedHTTP(tcpIngressIP, svcPort, loadBalancerCreateTimeout)
+		ginkgo.By("hitting the TCP service's LoadBalancer with no backends, no answer expected")
+		testNotReachableHTTP(tcpIngressIP, svcPort, loadBalancerLagTimeout)
```

Comment:
I think this is wrong. We expect load balancers to actively reject connections to services with no endpoints.

Reply:
This was added inin https://github.com/kubernetes/kubernetes/pull/74394, pinging @thockin  as the author,

I don't think we can enforce this behavior as this is asserting on public IP of the loadbalancer, so it is loadbalancer specific that is completely outside of kubernetes, per example, I'm doing this change because AFAIK haproxy can not implement this behavior, at least I could not find how

Reply:
I don't want the test just to pass, I want the test to assert on the behavior we want to define, and understanding if this is a behavior is reasonable so we don't have XX% of implementations non-conformant.

I also think that this reject behavior is only implemented with google l4 loadbalancers, AFAIK the L7 loadbalancers will send 50x errors https://cloud.google.com/knowledge/kb/502-503-errors-on-external-http-s-load-balancer-000006416 , @bowei can you confirm?

Reply:
Wait, this test is `e2eskipper.SkipUnlessProviderIs("gce", "gke", "aws")`. As written, it shouldn't be running with the kind CloudProvider anyway. We must not be running/configuring the e2e binary correctly.

(So we currently have no information about whether this behavior is widely supported, or just gce/aws. Fixing https://github.com/kubernetes/kubernetes/issues/123714 will let us figure that out...)

Reply:
> Wait, this test is e2eskipper.SkipUnlessProviderIs("gce", "gke", "aws"). As written, it shouldn't be running with the kind CloudProvider anyway. We must not be running/configuring the e2e binary correctly.

heh, I totally forget about it, in order to run the tests we need to set `  '--provider=aws`. to deal with some of the shenanigans of the providers


> (So we currently have no information about whether this behavior is widely supported, or just gce/aws

also discussed here https://github.com/kubernetes/kubernetes/issues/121371

Reply:
> heh, I totally forget about it, in order to run the tests we need to set `--provider=aws`

but we shouldn't do that; we should set `--provider=kind`. In which case this test won't get run and we can avoid thinking about it for now so you should remove this change from the PR

Reply:
the tests are expected to be removed https://github.com/kubernetes/kubernetes/issues/122828
 so we can not add new ones

I remember now that I used aws because it was null https://github.com/kubernetes/kubernetes/blob/b722d017a34b300a2284b890448e5a605f21d01e/test/e2e/framework/providers/aws/aws.go

gce adds a lot of stuff per example https://github.com/kubernetes/kubernetes/tree/b722d017a34b300a2284b890448e5a605f21d01e/test/e2e/framework/providers/gce

Reply:
I *think* I agree with Antonio here.  I don't think we need to assert rejection - it seems over-specific.

Reply:
> the tests are expected to be removed #122828
> so we can not add new ones

(I assume "new ones" meaning "new providers".)

Ah. I didn't realize that `SkipUnlessProviderIs` was going away... I thought the "provider" here was just an arbitrary string passed as an argument to `e2e.test`

The list of tests in #122828 seems to be wrong (since it includes this one, which is not provider-specific; it's just that we're incorrectly limiting it to certain providers, #123714).

Reply:
> I _think_ I agree with Antonio here. I don't think we need to assert rejection - it seems over-specific.

It's _consistent_. The behavior was changed in #74394, with the note "ICMP reject services with no endpoints through LBs. We already reject all other cases."; There had been a bunch of discussion in #48719 and everyone seemed to agree that it didn't make a lot of sense that we rejected connections to clusterIPs and externalIPs and NodePorts when there were no endpoints, but we didn't reject connections to LB IPs when there were no endpoints.

Though in this case, we could say "LBs _should_ do that, for consistency, but we don't _require_ that they do that".

----

Anyway, this behavior was specifically added to fix #48719, where the problem was that bad conntrack entries were being created on the nodes themselves. So the important part of the fix was the behavior _on the nodes_, not the behavior through the cloud LoadBalancer. So we could keep the kube-proxy changes from #74394 and rewrite the e2e changes (to only test the on-node behavior, not using the LB) and #48719 would remain fixed.

Alternatively, we could revert #74394 and fix #48719 differently by doing bad-conntrack-entry-cleanup for TCP blackhole entries in this case, like we do for UDP. But it seems like the main reason we didn't do it that way originally was because it seemed better to do it the #74394 way instead.

Reply:
filed #124338 to figure out the specific disposition of each of the LB e2e tests in the post-legacy-cloud-provider world

Reply:
> we could say "LBs should do that, for consistency, but we don't require that they do that".

This is what I meant to express, thanks.  I don't think we can dictate what LBs do when kube-proxy returns ICMP REJECT, so the test should assert "failed", whether that is fast (ICMP) or slow (DROP).  If GCE expects it to ICMP, that would be a GCE test, right?

Reply:
ping @danwinship , are we ok with this change?

Reply:
I'm OK with not asserting on the LB behavior, but as per the investigation above, there is an important node-level behavior that we were implicitly checking as well; if something breaks our `-j REJECT` rules in the future and the node goes back to silently dropping traffic in this case, then the e2e test as you've modified it here would still pass, but we'd start getting #48719 again.

Ideally we'd have a test of the specific behavior of #48719, but it relies on a race condition between the LB endpoints and the kube-proxy endpoints, so I'm not sure we can trigger it reliably. (We have this problem with testing the edge cases of eTP Local and LoadBalancerSourceRanges too... maybe we need to add a feature to cloud-provider-kind that allows us to trigger kube-proxy/cloud race conditions intentionally?)

We could try connecting to the LB IP from a pod so that we hit the short-circuit, but I'm not sure that #48719 actually triggers in that case...

Alternatively, we could just trust the results of the packet-flow tests and assume that if they say that we're rejecting packets, then we really are rejecting packets.

Reply:
the loadbalancer in cloud-provider-kind is forwarding to the nodes nodeports, is not forwarding to the clusterIP, you can test the reject by hitting the nodeports directly ... what rules are you wanting to validate exactly that send rejects without endpoints, the ClusterIP ones, the NodePort ones or the LoadBalancer ones (shortcut)?

Reply:
#48719 was about the LB IP, not NodePorts (which were already being rejected at that point).

I think we can just say that [the packet flow unit test](https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/proxy/iptables/proxier_test.go#L1879-L1885) is enough to satisfy us that, yes, we are rejecting connections to the LB IP when there are no endpoints.